### PR TITLE
Refactor decorated elements

### PR DESCRIPTION
**Pull request by yann-eugone**
There is more decorator elements every day.

The problem is that it is required to the developer to known how to init the decorated element. And thus replicate all the logic in every decorator.

This PR introduces an abstract class that will cover this requirement for almost every decorator.

### CODE REVIEW

File: src/batch/src/Job/Item/Processor/ChainProcessor.php [hunk 1]
```diff
@@ -59,10 +37,8 @@ public function process($item)
     /**
      * @inheritDoc
      */
-    public function flush(): void
+    protected function getDecoratedElements(): iterable
     {
-        foreach ($this->processors as $processor) {
-            $this->flushElement($processor);
-        }
+        return $this->processors;
```
Comment (J-Ben87) [inline src/batch/src/Job/Item/Processor/ChainProcessor.php:42]: Why not `yield from $this->processors;`?

File: src/batch/src/Job/Item/Reader/SequenceReader.php [hunk 2]
```diff
@@ -57,12 +35,10 @@ public function read(): iterable
...
-     * @inheritDoc
+     * @inheritdoc
      */
-    public function flush(): void
+    protected function getDecoratedElements(): iterable
     {
-        foreach ($this->readers as $reader) {
-            $this->flushElement($reader);
-        }
+        return $this->readers;
```
Comment (J-Ben87) [inline src/batch/src/Job/Item/Reader/SequenceReader.php:42]: Why not `yield from $this->readers;`?

File: src/batch/src/Job/Item/Writer/ChainWriter.php [hunk 3]
```diff
@@ -55,12 +33,10 @@ public function write(iterable $items): void
...
-     * @inheritDoc
+     * @inheritdoc
      */
-    public function flush(): void
+    protected function getDecoratedElements(): iterable
     {
-        foreach ($this->writers as $writer) {
-            $this->flushElement($writer);
-        }
+        return $this->writers;
```
Comment (J-Ben87) [inline src/batch/src/Job/Item/Writer/ChainWriter.php:40]: Why not `yield from $this->writers;`?

File: src/batch/src/Job/Item/AbstractElementDecorator.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * - {@see JobExecutionAwareInterface}
+ * - {@see SummaryAwareInterface}
+ * - {@see JobParametersAwareInterface}
+ */
+abstract class AbstractElementDecorator implements
+    InitializableInterface,
+    FlushableInterface,
+    JobExecutionAwareInterface
+{
+    use ElementConfiguratorTrait;
```
Comment (J-Ben87) [inline src/batch/src/Job/Item/AbstractElementDecorator.php:30]: Do you still need those traits now that there is an abstract class taking care of the logic? Couldn't you replace them by private methods?

File: src/batch/src/Job/Item/Processor/ChainProcessor.php (hunk 1 above)
Comment (yann-eugone) [reply src/batch/src/Job/Item/Processor/ChainProcessor.php:42]: Function accept any iterable, an array is iterable.

What will be the gain in wrapping this array into a Generator ?

File: src/batch/src/Job/Item/AbstractElementDecorator.php (hunk 4 above)
Comment (yann-eugone) [reply src/batch/src/Job/Item/AbstractElementDecorator.php:30]: Still required, this trait is also used in several classes cannot be child of this abstract class : 

- https://github.com/yokai-php/batch-src/blob/0.x/src/batch/src/Job/Item/ItemJob.php#L14
- https://github.com/yokai-php/batch-src/blob/0.x/src/batch/src/Job/Item/Processor/RoutingProcessor.php#L22
- https://github.com/yokai-php/batch-src/blob/0.x/src/batch/src/Job/Item/Writer/RoutingWriter.php#L22

File: src/batch/src/Job/Item/Processor/ChainProcessor.php (hunk 1 above)
Comment (J-Ben87) [reply src/batch/src/Job/Item/Processor/ChainProcessor.php:42]: Won't it free memory? I guess the gain is not huge here, but any performance improvement is good :slightly_smiling_face:

Comment (yann-eugone) [reply src/batch/src/Job/Item/Processor/ChainProcessor.php:42]: I don't think so, processors is already an array in use, so it has to be in memory.

Generator are only relevant when ALL the data flow is using an iterator.

Comment (J-Ben87) [reply src/batch/src/Job/Item/Processor/ChainProcessor.php:42]: All right then :slightly_smiling_face:

**Review by J-Ben87 — APPROVED** [review]

**Review by J-Ben87 — APPROVED** [review]

**Review by a-flaus — APPROVED** [review]

**Pull request merged**